zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/MigrationSnapshotDirector.java

```diff
@@ -0,0 +1,164 @@
+import io.camunda.zeebe.util.health.HealthMonitorable;
+import io.camunda.zeebe.util.health.HealthReport;
+import java.io.IOException;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MigrationSnapshotDirector implements HealthMonitorable, CloseableSilently {
+  public static final String COMPONENT_NAME = "MigrationSnapshotDirector";
+  private static final Logger LOG = LoggerFactory.getLogger(MigrationSnapshotDirector.class);
+
+  private volatile boolean snapshotTaken = false;
+  private ScheduledTimer runningSnapshot = null;
+  private final ConcurrentMap<FailureListener, Boolean> listeners = new ConcurrentHashMap<>();
+  private final AsyncSnapshotDirector snapshotDirector;
+
+  private volatile HealthReport healthReport;
+  private final Duration scheduleDelay;
+  private final ConcurrencyControl control;
+  private final HealthMonitor healthMonitor;
+
+  public MigrationSnapshotDirector(
+      final AsyncSnapshotDirector asyncSnapshotDirector,
+      final ConcurrencyControl control,
+      final Duration scheduleDelay,
+      final HealthMonitor healthMonitor) {
+    snapshotDirector = asyncSnapshotDirector;
+    this.control = control;
+    this.scheduleDelay = scheduleDelay;
+    this.healthMonitor = healthMonitor;
+    healthReport = snapshotNotTaken();
+    healthMonitor.registerComponent(this);
+    LOG.debug("Initialized migration snapshot director. Scheduling snapshot");
+    scheduleSnapshot();
+  }
+
+  @Override
+  public void close() {
+    cancelScheduledSnapshot();
+    healthMonitor.removeComponent(this);
+  }
+
+  public void scheduleSnapshot() {
+    if (!snapshotTaken) {
+      LOG.debug("Scheduling snapshot for migration.");
+      forceSnapshotUntilSuccessful();
+    }
+  }
+
+  public void cancelScheduledSnapshot() {
+    if (runningSnapshot != null) {
+      runningSnapshot.cancel();
+      runningSnapshot = null;
+    }
+  }
+
+  @VisibleForTesting
+  public boolean isSnapshotTaken() {
+    return snapshotTaken;
+  }
+
+  @Override
+  public String componentName() {
+    return COMPONENT_NAME;
+  }
+
+  @Override
+  public HealthReport getHealthReport() {
+    return healthReport;
+  }
+
+  @Override
+  public void addFailureListener(final FailureListener failureListener) {
+    listeners.put(failureListener, Boolean.TRUE);
+  }
+
+  @Override
+  public void removeFailureListener(final FailureListener failureListener) {
+    listeners.remove(failureListener);
+  }
+
+  private void forceSnapshotUntilSuccessful() {
+    forceSnapshotUntilSuccessful(null);
+  }
+
+  private void forceSnapshotUntilSuccessful(final Throwable error) {
+    var shouldBeRetried = !snapshotTaken && runningSnapshot == null;
+    if (error != null) {
+      shouldBeRetried &=
```

Comment:
🔧 I would extract this in a helper, e.g. isRecoverableError, and just call that in the declaration of shouldBeRetried.